File: src/main/java/com/apschulewitz/resdb/common/model/entity/Title.java
package com.apschulewitz.resdb.common.model.entity;

import com.apschulewitz.resdb.common.model.VersionableDataEntity;
import lombok.Builder;
import lombok.Data;
import lombok.experimental.Tolerate;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Table;
import javax.persistence.UniqueConstraint;
import java.time.ZonedDateTime;

/**
 * @author adrian
 */
@Data
@Builder
@Entity
@Table(name = "resdb_title", uniqueConstraints = @UniqueConstraint(columnNames = {"title"}))
//@Audited
public class Title implements VersionableDataEntity<Long> {

  private static final long serialVersionUID = -1411917831665423698L;

  @Id
  @GeneratedValue(strategy = GenerationType.AUTO)
  private Long id;

  @Column(nullable = false, length = 30)
  private String title;

  @Column(length = 250)
  private String description;

  @Column(nullable = false, length = 1) //, name = "applies_to"
  private Gender appliesTo;

  @Column(nullable = false, length = 1)
  private TitleType titleType;

  @Column
  private VersionStatus status;

  @Column(nullable = false, length = 20)
  private String createdBy;

  @Column
  private String updatedBy;

  @Column(name = "version_no")
  private Long versionNumber;

  @Column(name = "last_updated", length = 20)
  private ZonedDateTime lastUpdated;

  private transient DataOperation operation;

  @Tolerate
  public Title() {

  }
}

File: src/app/model/entity/title.ts
import { TitleType } from './title-type';
import {DataStatus} from '../../core/model/data-status';
import {DataAction} from '../../core/model/data-action';
import {Gender} from './gender';
import {IAuditedDataType} from './interfaces/audited-data-type';
import {CodeLabel} from './code-label';

export class Title implements IAuditedDataType {

  id: number;
  title: string;
  description: string;
  titleType: TitleType;
  appliesTo: Gender; // CodeLabel;
  status: DataStatus;
  createdBy: string;
  updatedBy: string;
  lastUpdated: Date;
  selected: boolean;
  action: DataAction;
  isDataChanged: boolean;

  constructor() {
    this.isDataChanged = false;
    this.selected = false;
    this.id = 0;
    this.title = '';
    this.action = DataAction.Add;
    this.status = DataStatus.New;
  }

  public toString(): string {
    return 'Title: id=' + this.id + ' title=' + this.title + ' description=' + this.description + ' titleType=' + this.titleType +
      ' appliesTo=' + this.appliesTo +
      ' status=' + this.status + ' createdBy=' + this.createdBy +
      ' updatedBy=' + this.updatedBy + ' lastUpdated=' + this.lastUpdated + ' selected=' + this.selected + ' action=' + this.action +
      ' isDataChanged=' + this.isDataChanged;
  }
}
